src/common/cockpitframe.c

```diff
@@ -139,103 +144,155 @@ cockpit_frame_write (int fd,
+              *was_eof = true;
+              return true;
             }
-        }
 
-      if (size < 0)
-        {
-          errn = EBADMSG;
-          goto out;
+          errno = EBADMSG;
+          return false;
         }
+
+      offset += n;
     }
 
-  if (output)
+  if (was_eof)
+    *was_eof = false;
+
+  return true;
+}
+
+ssize_t
+cockpit_frame_read (int fd,
+                    unsigned char **output)
+{
+  /* We first need to read the size of the frame, followed by the
+   * content of the frame.  We want to do this efficiently as possible,
+   * while avoiding to read() more than the frame (since we can't put
+   * bytes back).  MSG_PEEK is also not always available, since we're
+   * often talking to a pipe.
+   *
+   * Fortunately, we have a reasonable approach for this.
+   *
+   * Empty frames are invalid (cockpit_frame_parse() rejects size == 0),
+   * so the smallest possible frame has a length of at least 3: the
+   * single-digit size, the newline, then the single byte of body.
+   * Therefore it's always safe to read 3 bytes ("the initial read").
+   *
+   * Conveniently, reading three bytes is also always enough to tell us
+   * how many bytes it's safe to read in order to determine the size of
+   * the frame:
+   *
+   *   - if we read a digit or two, followed by a newline, then we
+   *     already know the size of the entire frame
+   *
+   *   - if we read three digits, we know that the frame body is at
+   *     least 100 bytes long.  Since the maximum size of the length
+   *     field is 7 (+1 for newline), we can safely read this entire
+   *     amount.
+   *
+   *   - if we get something other than digits or newlines, it's an
+   *     error
+   */
+  size_t n_read = 3;
+  unsigned char headerbuf[MAX_FRAME_SIZE_BYTES + 1];
+  bool eof;
+
+  if (!read_exactly (fd, headerbuf, n_read, &eof))
+    return -1;
+
+  if (eof)
     {
-      *output = buf;
-      buf = NULL;
+      if (output)
+        *output = NULL;
+      return 0;
     }
-  ret = size;
 
-out:
-  free (buf);
-  if (ret < 0)
-    errno = errn;
-  return ret;
+  size_t n_consumed;
+  ssize_t size = cockpit_frame_parse (headerbuf, n_read, &n_consumed);
+  if (size == 0)
+    {
+      /* cockpit_frame_parse() asked to read more data.  As explained
+       * above, it's safe to read the rest of the buffer now (5 bytes).
+       */
+      if (!read_exactly (fd, headerbuf + n_read, sizeof headerbuf - n_read, NULL))
+        return -1;
+
+      n_read = sizeof headerbuf;
+      size = cockpit_frame_parse (headerbuf, n_read, &n_consumed);
+    }
+
+  if (size == -1)
+    return -1; /* EBADMSG */
+
+  /* We now have size equal to the number of bytes we need to return. */
+  unsigned char *buffer = malloc (size);
+  if (buffer == NULL)
+    return -1; /* ENOMEM */
+
+  /* Copy the non-consumed bytes from the header (might be zero) */
+  size_t bytes_from_header = n_read - n_consumed;
+  memcpy (buffer, headerbuf + n_consumed, n_read - n_consumed);
```

Comment:
nitpicking, but I'd find s/n_read - n_consumed/bytes_from_header/ slightly easier to read. It also matches up better with the read_exactly() below.

Reply:
Thanks for this suggestion!  This was probably also my original intention but I clearly missed it.